hibike/hibike_message.py

```diff
@@ -112,31 +131,195 @@ def checksum(data):
+      device_id to properly construct the message.
+      
+      device_id         - a device type id (not uid).
+      params_and_values - an iterable of param (name, value) tuples
+  """
+  params_and_values = sorted(params_and_values, key=lambda x: paramMap[device_id][x[0]][0])
+  params = [param[0] for param in params_and_values]
+  paramNums = [paramMap[device_id][name][0] for name in params]
+  paramT = [paramMap[device_id][name][1] for name in params]
+  values = [param[1] for param in params_and_values]
+  entries = [1 << num for num in paramNums]
+
+  tot = 0
+  for i in range(len(entries)):
+    tot = tot ^ entries[i]
+  typeString = '<H'
+  for t in paramT:
+    typeString += paramTypes[t]
+  temp_payload = struct.pack(typeString, tot, *values)
   payload = bytearray(temp_payload)
-  message = HibikeMessage(messageTypes["DeviceUpdate"], payload)
+  message = HibikeMessage(messageTypes["DeviceWrite"], payload)
   return message
 
-def make_device_status(param):
-  """ Makes and returns DeviceStatus message."""
-  temp_payload = struct.pack('<B', param)
+def old_decode_device_write(device_id, message):
+  messageID = message.getmessageID()
+  payload = message.getPayload()
+  messageT = "DeviceWrite"
+  paramB = bin(payload[0])
+  paramNames = decode_params(device_id, paramB)
+  paramT = [paramMap[device_id][name][1] for name in paramNames]
+  typeString = '<H'
+  for t in paramT:
+    typeString += paramTypes[t]
+  tot_values = struct.unpack(typeString, payload)
+  params_and_values = []
+  for i in range(len(paramNames)):
+    param_and_values.append((paramNames[i], tot_values[i+1]))
+  return params_and_values
+
+def decode_device_write(msg, device_id):
+  assert msg.getmessageID() == messageTypes["DeviceWrite"]
+  payload = msg.getPayload()
+  assert len(payload) >= 2
+  params,  =  struct.unpack("<H", payload[:2])
+  params = decode_params(device_id, params)
+  struct_format = "<"
+  for param in params:
+    struct_format += paramTypes[paramMap[device_id][param][1]]
+  values = struct.unpack(struct_format, payload[2:])
+  return list(zip(params, values))
+
+
+
+
+
+
+
+def make_device_data(device_id, params_and_values):
+  """ Makes and returns DeviceData message.
+      If all the params cannot fit, it will fill as many as it can.
+
+      looks up config data about the specified 
+      device_id to properly construct the message.
+      
+      device_id         - a device type id (not uid).
+      params_and_values - an iterable of param (name, value) tuples
+  """
+  params = [param_tuple[0] for param_tuple in params_and_values]
+  paramNums = [paramMap[device_id][name][0] for name in params]
+  entries = [1 << num for num in paramNums]
+
+  tot = 0
+  for i in range(len(entries)):
+    tot = tot ^ entries[i]
+  
+  paramT = [paramMap[device_id][name][1] for name in params]
+  values = [param_tuple[1] for param_tuple in params_and_values]
+
+  typeString = '<H'
+  for type in paramT:
+    typeString += paramTypes[type]
```

Comment:
Construction of a type string from param types should be abstracted into a function.